### DISCUSSION TITLE
DDD

### DISCUSSION THREAD

**Discussion by ivansglazunov** [deep-foundation/dev#39] [Posts]
[Data Driven Development](https://en.wikipedia.org/wiki/Data-driven_programming) парадигма применяется повсюду, и нигде.

Правильный DDD это:
- Все обращения в системе происходят только к единому хранилищу данных.
- К runtime/microservice/API обращения происходят только от единого хранилища данных.
- Права на выполнение тех или иных действия описаны в одном единственном месте декларативным кодом, а императивный не осведомлен о их существовании.
- Все runtime/microservice/API обращаются только к единому хранилищу данных для интеграции своего функционала или к внешним сервисам для реализации собственной работы.
- Императивный функционал не осведомлен о бизнес логике которой он способствует, и его участие в ней можно изменить путем модификации данных в хранилище.
- Изменения в хранилище данных приводят к обновлению зависимых поведений.

Примеры
- Kubernetes отражает декларативную иммутабельную конфигурацию отказав в кастомизации его императивного источника, отсутствие очевидного одинакового жизненного цикла у всех сущностей системы делает DDD здесь сугубо конфигурирующим свойством, не способным задавать способ зависимости сущностей друг с другом
- React есть только нисходящий поток иммутабельных данных которые могут исходить из любого императивного источника, Angular/Vue и прочие веб фреймворки так-же подошли близко к DDD, за счет открытой разработки, наибольшему рынку связанных разработчиков
- Микросервисная архитектура звучит как мейнстрим и постоянно работает с некоторыми хранилищами данных, но в этой культуре нет ничего от DDD, если к API происходят обращения не только от хранилища данных, и API знает что-то о бизнес логике кроме своей непосредственной функции
- Микроконтроллеры уходят с головой в протоколы связи и языки, DDD почти не развит, высокий порог входа мешает легко зайти на этот рынок разработчиком из иных культур разработки

**Comment by Menzorg** [comment]
И в статье говорится о data driven programming, мы намеренно расширяем программирование до разработки, чтобы покрыть большую область применения парадигмы? В таком случае devOps при применении IaC подхода в декларативном стиле тоже форма DDD? Когда мы описываем в репозитории состояние требуемое нам, и CD исходя из данных - производит изменения в архитектуре.
https://en.wikipedia.org/wiki/Infrastructure_as_code

**Reply by Menzorg** [reply]
И это можно добавить в список примеров

**Reply by ivansglazunov** [reply]
Да
